spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java

```diff
@@ -150,8 +167,36 @@ else if (reply instanceof AbstractIntegrationMessageBuilder) {
 			}
 		}
 
-		Message<?> replyMessage = createOutputMessage(reply, requestHeaders);
-		sendOutput(replyMessage, replyChannel);
+		if (this.asyncReplySupported && reply instanceof ListenableFuture<?>) {
+			ListenableFuture<?> future = (ListenableFuture<?>) reply;
+			final Object theReplyChannel = replyChannel;
+			future.addCallback(new ListenableFutureCallback<Object>() {
+
+				@Override
+				public void onSuccess(Object result) {
+					sendOutput(createOutputMessage(result, requestHeaders), theReplyChannel, false);
+				}
+
+				@Override
+				public void onFailure(Throwable ex) {
+					Object errorChannel = requestHeaders.getErrorChannel();
+					if (errorChannel == null) {
+						throw new MessageHandlingException(requestMessage,
+								"Async exception received and no 'errorChannel' header exists; cannot route"
+								+ "exception to caller", ex);
```

Comment:
A whitespace between `route` and `exception` words.
Otherwise LGTM and I can merge, if you don't have anything else to utilize here.